### ISSUE TITLE
Need help to use ng-logger using npm link method

### ISSUE THREAD

**Issue by JigneshRaval** [noemi-salaun/ng-logger#9]
Hi @noemi-salaun ,

I am really thankful to you for providing such a simple example of Angular Module.

I need you guidance for one scenario.

Node v7.3
Angular 4.0.

I have tried to following way.
1) I installed this module using "npm install --save @nsalaun/ng-logger" and imported it in my angular 4 app as written in Readme file and was working fine.

2) I downloaded your folder as zip file and extracted it somewhere in my local drive and did "npm install" then ran "npm link" command inside from that extracted folder and in my angular app i ran "npm link ng-logger"
so it attached with my angular app. 
I imported it using "import { NgLoggerModule, Level } from 'ng-logger';"
and in @ngModule's import section I added "NgLoggerModule.forRoot(Level.LOG)"
but I am getting following error.

`ERROR in Error encountered resolving symbol values statically. Calling function 'OpaqueToken', function calls are not supported. Consider re
placing the function or lambda with a reference to an exported function, resolving symbol LOGGER_LEVEL in C:/jr/__UPointNext/project/alight-
app/node_modules/alight-test2/src/logger.service.d.ts, resolving symbol NgLoggerModule.forRoot in C:/jr/__UPointNext/project/alight-app/node
_modules/alight-test2/src/ng-logger.module.d.ts, resolving symbol AppModule in C:/jr/__UPointNext/project/alight-app/src/app/app.module.ts,
resolving symbol AppModule in C:/jr/__UPointNext/project/alight-app/src/app/app.module.ts`

I am using Angular CLI 1.1.3 in my app in which I am linking your module.

``"dependencies": {
    "@angular/animations": "^4.0.0",
    "@angular/common": "^4.0.0",
    "@angular/compiler": "^4.0.0",
    "@angular/core": "^4.0.0",
    "@angular/forms": "^4.0.0",
    "@angular/http": "^4.0.0",
    "@angular/platform-browser": "^4.0.0",
    "@angular/platform-browser-dynamic": "^4.0.0",
    "@angular/router": "^4.0.0",
    "@nsalaun/ng-logger": "^2.0.1",
    "core-js": "^2.4.1",
    "rxjs": "^5.1.0",
    "zone.js": "^0.8.4"
  },
  "devDependencies": {
    "@angular/cli": "1.1.3",
    "@angular/compiler-cli": "^4.0.0",
    "@angular/language-service": "^4.0.0",
    "@types/jasmine": "2.5.45",
    "@types/node": "~6.0.60",
    "codelyzer": "~3.0.1",
    "jasmine-core": "~2.6.2",
    "jasmine-spec-reporter": "~4.1.0",
    "karma": "~1.7.0",
    "karma-chrome-launcher": "~2.1.1",
    "karma-cli": "~1.0.1",
    "karma-coverage-istanbul-reporter": "^1.2.1",
    "karma-jasmine": "~1.1.0",
    "karma-jasmine-html-reporter": "^0.2.2",
    "protractor": "~5.1.2",
    "ts-node": "~3.0.4",
    "tslint": "~5.3.2",
    "typescript": "~2.3.3"
  }``

could you please guide where I am lacking. 
I also tried it in non CLI Angular 4 project ( Manually setup of angular 4 project using webpack) but in that it gives following error.

Uncaught Error: Can't resolve all parameters for Logger: (?).
    at syntaxError (compiler.es5.js:1540)
    at CompileMetadataResolver._getDependenciesMetadata (compiler.es5.js:14877)
    at CompileMetadataResolver._getTypeMetadata (compiler.es5.js:14745)
    at CompileMetadataResolver._getInjectableMetadata (compiler.es5.js:14731)
    at CompileMetadataResolver.getProviderMetadata (compiler.es5.js:15021)
    at compiler.es5.js:14950
    at Array.forEach (<anonymous>)
    at CompileMetadataResolver._getProvidersMetadata (compiler.es5.js:14911)
    at compiler.es5.js:14486
    at Array.forEach (<anonymous>)

**Comment by noemi-salaun** [comment]
@JigneshRaval 
I have never used npm link but I can try to help you.
What folder did you download ? The GitHub zip with the sources ?

**Comment by JigneshRaval** [comment]
Thanks @noemi-salaun ,

Yes, I have downloaded it from this repo using download button available on this page 
https://github.com/noemi-salaun/ng-logger

**Comment by noemi-salaun** [comment]
You need to build the logger, the same way I did before publishing it on NPM.

`npm install` then `npm run bundle` should do the job.

**Comment by JigneshRaval** [comment]
Sure, I will try that.

Thank you,

**Comment by JigneshRaval** [comment]
Hi @noemi-salaun ,

Can you please do me a favour.

Can you please provide one small example which just do console.log("Hi"). using following way.

1) download your github repo of ng-logger
2) extract that zip file and from that do "npm link"
- this will create a global link in users/appdata/roaming/npm folder.
3) create one angular project using Angular CLI v1.1.3
4) go to that newly created CLI app and execute command prompt and run "npm link ng-logger"
- after linking ng-logger will be available in our new app's node_modules folder with it's own nested "node-modules" folder
5) in App.module.ts import ng-logger as we do normally like import { test } from ng-logger
and just run app using ng serve
This time in ng-logger module, we will not export all the files but only a simple function with console.log

This will be your great help.

Thanks,
Jignesh

**Comment by noemi-salaun** [comment]
It looks like `npm link` doesn't handle the linked package dependencies. Instead of `npm link @nsalaun/ng-logger`, try using `npm install --save /path/to/the/folder/ng-logger`

[`npm install`](https://docs.npmjs.com/cli/install) can install from multiple sources.

> a folder containing a program described by a package.json file

**Comment by JigneshRaval** [comment]
Do I need to run this command from my another angular CLI app?

npm install --save /path/to/the/folder/ng-logger

**Comment by noemi-salaun** [comment]
Yes, if you want to use the logger in this app. But I don't understand for what use case you need to download the sources and build the lib yourself.

**Comment by JigneshRaval** [comment]
Actually we are building one app.
- In which we will have two separate angular cli app. 
(1) Core App (2) Normal App
- in Core app we will put all our common components , services and other generic code which will export to another app.
- Then in second app we will import required components, services or any angular modules like pipes, Directive from that Core App module.
- For that we are not able to publish our code to NPM because of policy.
- So we have to use npm link method. which serves a good way to link core app into second app without publishing to NPM/installing from NPM. It is same like virtual hosting of one app. Which we can link and use in any other app.
- We created this way but it was working for angular version 2.3.1 and cli version 1.0 beta
- But after upgrading to 4 it is not working.

Please suggest.

Thanks,

**Comment by noemi-salaun** [comment]
I'am not sure I fully understand your project, but why don't you create your Core App, define some dependencies with `package.json`, and then in the Normal App, you `npm install --save /path/to/core-app`. 

`npm link` is not designed to resolve dependencies, you cannot use it to perform a proper install.

**Comment by noemi-salaun** [comment]
Your application could not just be a library exposing a module, just as my library exposes NgLoggerModule?

So your CoreModule project has a dependency on my logger with `npm install --save @nsalaun/ng-logger`, and your MainApp project has a dependency on your CoreModule project, with `npm install --save /path/to/core-project`.

In your CoreModule, you can use :

```
@NgModule({
  imports     : [
    // ...
    NgLoggerModule,
    // ....
  ]
})
export class CoreModule {
}
```

and in your AppModule :

```
@NgModule({
  declarations: [
    AppComponent,
  ],
  imports     : [
    BrowserModule,

    NgLoggerModule.forRoot(Level.LOG),
    CoreModule,
    // CoreModule.forRoot() // If you want to provide some singletons services.
  ],
  providers   : [],
  bootstrap   : [AppComponent],
})
export class AppModule {
}
```

**Comment by JigneshRaval** [comment]
Thanks @noemi-salaun ,

I tried to do do `npm install --save /path/to/ng-logger` in my MainApp but it just installed following files only.
![image](https://user-images.githubusercontent.com/11232532/27682408-0a4f7a74-5ce0-11e7-8ada-d773529b555d.png)

**Comment by JigneshRaval** [comment]
Could you please guide that when I do `npm install --save @nsalaun/ng-logger` in MainApp then how it generates highlighted files shown in below screenshot?

while in your git repo only `*.d.ts` files available.

![image](https://user-images.githubusercontent.com/11232532/27682687-295c7dda-5ce1-11e7-969f-25d838326fe8.png)

**Comment by noemi-salaun** [comment]
I don't have *.d.ts file in my GitHub repo, only the sources, with the *.ts.
If you don't want to alter the code of the logger, you shouldn't have to clone the repo. Just use `npm install --save @nsalaun/ng-logger`.
The version published on NPM is already built and contains only the files you needed for compiling your app and running it.
You don't need my *.ts files and you don't need all the files used only for building my lib (karma.*, tsconfig.*, etc.)

What I understand of your needs is :

**MainApp**  ---requires--->  **CoreApp**  ---requires---> **ng-logger**

But in this case, what you call CoreApp, is juste another Angular libraries, just like my ng-logger, or like any other Angular libraries. It's not an app, it's just a collection of Services, Components, Pipes, ... provided in a coherent module. Your CoreApp needs to be bundle like any other Angular libs.

I think you can follow the article you found :
https://medium.com/@OCombe/how-to-publish-a-library-for-angular-2-on-npm-5f48cdabf435

If you don't want to publish it on NPM, it's fine. You MainApp can depends on your CoreApp with `npm install --save /path/to/core/project/folder`. But the CoreApp must be already bundle like a lib published on NPM.

I can not help you much more, I have just built a logger to be used by the Angular community. If you want to build a complex application with many module dependencies and more, you should follow the appropriates tutorials.

**Comment by JigneshRaval** [comment]
Hi @noemi-salaun ,

You are right, `npm link` doesn't work properly. I tried to create simple plugin and that was working fine using '`npm install --save /path/to/my/plugin/` but the same thing is not working while using `npm link`.

It seems that there is some path related issues in `npm link`. when I link one module or app into another app then internal module path is not resolved properly.

I have put my example here : https://github.com/JigneshRaval/link-test

I am very thankful to you for your great help.

Thanks again,
Jignesh

**Comment by JigneshRaval** [comment]
Hi @noemi-salaun ,

After so much R&D, finally I tried to resolve issue in `npm link`. That was issue of webpack which was not resolving dependacies properly.

I have created one example and put it here : https://github.com/JigneshRaval/linked-modules-example

MyModule-A will be linked using `npm link` and then in MyModule-B we can do `npm link MyModule-A`.

Then run `npm run start` and `node server.js` from Module A.

Thanks,
Jignesh Raval
